File: scripts/bootstrap_db.py
from __future__ import annotations

import os
import re
import struct
import sys
from pathlib import Path

import pyodbc
from azure.identity import AzureCliCredential

ACCESS_TOKEN_ATTRIBUTE = 1256


def connect() -> pyodbc.Connection:
    token = AzureCliCredential().get_token(
        "https://database.windows.net/.default"
    ).token
    token_bytes = token.encode("utf-16-le")
    packed = struct.pack(f"<I{len(token_bytes)}s", len(token_bytes), token_bytes)
    connection_string = (
        "Driver={ODBC Driver 18 for SQL Server};"
        f"Server=tcp:{os.environ['SQL_SERVER']},1433;"
        f"Database={os.environ['SQL_DATABASE']};"
        "Encrypt=yes;TrustServerCertificate=no;Connection Timeout=30;"
    )
    return pyodbc.connect(
        connection_string,
        attrs_before={ACCESS_TOKEN_ATTRIBUTE: packed},
        autocommit=True,
    )


def batches(sql_text: str):
    return [
        batch.strip()
        for batch in re.split(r"^\s*GO\s*$", sql_text, flags=re.MULTILINE | re.IGNORECASE)
        if batch.strip()
    ]


def execute_file(connection: pyodbc.Connection, path: Path) -> None:
    text = path.read_text(encoding="utf-8")
    if path.name == "grants.sql":
        text = re.sub(r"^:setvar.*$", "", text, flags=re.MULTILINE)
        text = text.replace(
            "$(WebIdentityName)", os.environ["AZURE_CONTAINER_APP_NAME"]
        )
        text = text.replace(
            "$(IngestIdentityName)", os.environ["AZURE_INGEST_JOB_NAME"]
        )
        text = text.replace(
            "$(WebIdentityClientId)",
            os.environ["AZURE_CONTAINER_APP_CLIENT_ID"],
        )
        text = text.replace(
            "$(IngestIdentityClientId)",
            os.environ["AZURE_INGEST_JOB_CLIENT_ID"],
        )
    for index, batch in enumerate(batches(text), start=1):
        try:
            connection.cursor().execute(batch)
        except Exception as error:
            raise RuntimeError(f"{path} batch {index} failed") from error


def main() -> None:
    paths = [Path(argument) for argument in sys.argv[1:]]
    if not paths:
        raise SystemExit("Usage: bootstrap_db.py <sql-file> [<sql-file>...]")
    with connect() as connection:
        for path in paths:
            execute_file(connection, path)
            print(f"Applied {path}")


if __name__ == "__main__":
    main()

File: scripts/bootstrap_db.mjs
import { AzureCliCredential } from "@azure/identity";
import sql from "mssql";
import { readFile } from "node:fs/promises";
import { basename } from "node:path";

function required(name) {
  const value = process.env[name];
  if (!value) throw new Error(`${name} is required`);
  return value;
}

function batches(text) {
  return text
    .split(/^\s*GO\s*$/gim)
    .map((batch) => batch.trim())
    .filter(Boolean);
}

async function fileBatches(path) {
  let text = await readFile(path, "utf8");
  if (basename(path).toLowerCase() === "grants.sql") {
    text = text.replace(/^:setvar.*$/gim, "");
    text = text
      .replaceAll("$(WebIdentityName)", required("AZURE_CONTAINER_APP_NAME"))
      .replaceAll("$(IngestIdentityName)", required("AZURE_INGEST_JOB_NAME"))
      .replaceAll(
        "$(WebIdentityClientId)",
        required("AZURE_CONTAINER_APP_CLIENT_ID")
      )
      .replaceAll(
        "$(IngestIdentityClientId)",
        required("AZURE_INGEST_JOB_CLIENT_ID")
      );
  }
  return batches(text);
}

async function main() {
  const paths = process.argv.slice(2);
  if (!paths.length) {
    throw new Error("Usage: node bootstrap_db.mjs <sql-file> [<sql-file>...]");
  }

  const credential = new AzureCliCredential();
  const accessToken = await credential.getToken(
    "https://database.windows.net/.default"
  );
  if (!accessToken) throw new Error("Could not acquire an Azure SQL access token.");

  const pool = await new sql.ConnectionPool({
    server: required("SQL_SERVER"),
    database: required("SQL_DATABASE"),
    port: 1433,
    connectionTimeout: 30_000,
    requestTimeout: 120_000,
    options: {
      encrypt: true,
      trustServerCertificate: false
    },
    authentication: {
      type: "azure-active-directory-access-token",
      options: { token: accessToken.token }
    }
  }).connect();

  try {
    for (const path of paths) {
      const statements = await fileBatches(path);
      for (const [index, statement] of statements.entries()) {
        try {
          await pool.request().batch(statement);
        } catch (error) {
          throw new Error(`${path} batch ${index + 1} failed`, { cause: error });
        }
      }
      console.log(`Applied ${path}`);
    }
  } finally {
    await pool.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
